File: c++/MergeSort.cpp
#include "stdafx.h"
#include "MergeSort.h"

using namespace std;

vector<int> mergeTwoSortedArray(vector<int> &A, vector<int> &B)
{
	vector<int> C;

	int idxA = 0, idxB = 0;
	while (A.size() > idxA && B.size() > idxB)
	{
		int a = A.at(idxA);
		int b = B.at(idxB);

		if (a <= b)
		{
			C.push_back(a);
			idxA++;
		}
		else
		{
			C.push_back(b);
			idxB++;
		}
	}

	for (int i = idxA; i < A.size(); i++)
	C.push_back(A.at(i));

	for (int i = idxB; i < B.size(); i++)
	C.push_back(B.at(i));

	return C;
}

vector<int> mergeSort(int arr[], int st, int ed)
{
	if (st == ed) {
		vector<int> ret;
		ret.push_back(arr[st]);
		return ret;
	}

	int m = st + floor((ed - st) / 2);

	vector<int> A = mergeSort(arr, st, m);
	vector<int> B = mergeSort(arr, m + 1, ed);
	return mergeTwoSortedArray(A, B);
}

MergeSort::MergeSort()
{
	int arr[] = { 7, 2, 5, 3, 7, 13, 1, 6 };
	vector<int> result = mergeSort(arr, 0, 7);

	for (vector<int>::iterator i = result.begin(); i < result.end() ; i++)
	{
		cout << *i << " ";
	}
}


MergeSort::~MergeSort()
{
}

File: java/src/MergeSort.java
import java.util.*;

public class MergeSort {

	private static List<Integer> mergeSort(List<Integer> arr, int st, int ed) {
		if (st == ed) {
			return Arrays.asList(arr.get(st));
		}
		
		int m = st + (int) Math.floor((ed - st) / 2);
		List<Integer> A = mergeSort(arr, st, m);
		List<Integer> B = mergeSort(arr, m + 1, ed);
		return merge(A, B);
	}
	
	private static List<Integer> merge(List<Integer> A, List<Integer> B) {
		List<Integer> C = new ArrayList<>();
		
		int idxA = 0, idxB = 0;
		
		while (A.size() > idxA && B.size() > idxB) {
			int a = A.get(idxA);
			int b = B.get(idxB);
			if (a <= b) {
				C.add(a);
				idxA++;
			} else {
				C.add(b);
				idxB++;
			}
		}
		
		for (int i = idxA ; i < A.size() ; i++)
			C.add(A.get(i));

		for (int i = idxB ; i < B.size() ; i++)
			C.add(B.get(i));
		
		return C;
	}
	 
	public static void main(String[] args) {
		List<Integer> arr = Arrays.asList(7,2,5,3,7,13,1,6);
		
		for (int i : mergeSort(arr, 0, arr.size() - 1)) {
			System.out.print(i + " ");
		}
	}
}
